### PR TITLE
Wip/vl/ltlfixes

### PR DESCRIPTION
**Pull request by Langenfeld**
fixes several LTL checker bugs.

### CODE REVIEW

File: trunk/examples/LTL/regression/c/LTLAutomizerCInline.xml.tmp [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<rundefinition>
```
Comment (danieldietsch) [inline trunk/examples/LTL/regression/c/LTLAutomizerCInline.xml.tmp]: Why is this file committed?

File: trunk/examples/LTL/regression/c/stepwise/X_smaller_safe.c [hunk 2]
```diff
@@ -1,5 +1,5 @@
 //#Safe
-//@ ltl invariant positive: [](AP(y == 1) ==> X(AP(x == 2) ));
+//@ ltl invariant positive: [](AP(y == 0) ==> X(AP(x == 2) ));
```
Comment (danieldietsch) [inline trunk/examples/LTL/regression/c/stepwise/X_smaller_safe.c]: Why this change? To avoid the issue with wrong ``Ultimate.init``?

File: trunk/examples/LTL/regression/c/vanilla/Bug_ProcedureHandling.c [hunk 3]
```diff
@@ -1,54 +1,54 @@
...
-				goto recheck;
-			}
-			StrategyInvalidateBuffer(); 
-		}
-	}
-        
- my_exit:
-	while(1) { int yyy;yyy=yyy;}
+//#Unsafe
+//@ ltl invariant positive: <>AP(istemp == 0) || []AP(A != 1);
```
Comment (danieldietsch) [inline trunk/examples/LTL/regression/c/vanilla/Bug_ProcedureHandling.c:2]: Why? The idea here was: either init sets istemp to 0 and then everything is allowed or its some other value which has to be different from 1.

The new property should always be safe b.c. of line 22, making the classification wrong.

File: trunk/source/BuchiProgramProduct/src/de/uni_freiburg/informatik/ultimate/buchiprogramproduct/ProductBacktranslator.java [hunk 4]
```diff
@@ -87,9 +87,13 @@ public ProductBacktranslator(final Class<? extends IIcfgTransition<IcfgLocation>
 				newBranchEncoders.add(oldBranchEncoders[i]);
 			}
 		}
-
-		return IcfgProgramExecution.create(newTrace, newValues,
-				newBranchEncoders.toArray(new Map[newBranchEncoders.size()]));
+		if (!newTrace.isEmpty()) {
```
Comment (danieldietsch) [inline trunk/source/BuchiProgramProduct/src/de/uni_freiburg/informatik/ultimate/buchiprogramproduct/ProductBacktranslator.java]: flip condition

File: trunk/source/BuchiProgramProduct/src/de/uni_freiburg/informatik/ultimate/buchiprogramproduct/ProductBacktranslator.java [hunk 5]
```diff
@@ -48,7 +48,7 @@
  *
  */
 public class ProductBacktranslator extends
-		DefaultTranslator<IIcfgTransition<IcfgLocation>, IIcfgTransition<IcfgLocation>, Term, Term, String, String> {
+DefaultTranslator<IIcfgTransition<IcfgLocation>, IIcfgTransition<IcfgLocation>, Term, Term, String, String> {
```
Comment (danieldietsch) [inline trunk/source/BuchiProgramProduct/src/de/uni_freiburg/informatik/ultimate/buchiprogramproduct/ProductBacktranslator.java]: Is this CC compatible?

File: trunk/source/BuchiProgramProduct/src/de/uni_freiburg/informatik/ultimate/buchiprogramproduct/productgenerator/ProductGenerator.java [hunk 6]
```diff
@@ -72,12 +72,14 @@
 import de.uni_freiburg.informatik.ultimate.plugins.generator.rcfgbuilder.cfg.Summary;
 
 /**
- * This class is implementing the Buchi program product, i.e. interleaving a BuchiAutomaton with the CFG.
+ * This class is implementing the Buchi program product, i.e. interleaving a
```
Comment (danieldietsch) [inline trunk/source/BuchiProgramProduct/src/de/uni_freiburg/informatik/ultimate/buchiprogramproduct/productgenerator/ProductGenerator.java]: Is this CC compatible? Do you have line length 120?

File: trunk/source/BuchiProgramProduct/src/de/uni_freiburg/informatik/ultimate/buchiprogramproduct/productgenerator/ProductGenerator.java [hunk 7]
```diff
@@ -176,28 +178,36 @@ private void collectRcfgLocations() {
...
 			unhandledLocations.remove(currentPoint);
 			mRCFGLocations.add(currentPoint);
+
 			for (final IcfgEdge p : currentPoint.getOutgoingEdges()) {
 				if (!mRCFGLocations.contains(p.getTarget()) && !unhandledLocations.contains(p.getTarget())) {
 					unhandledLocations.add((BoogieIcfgLocation) p.getTarget());
 				}
 			}
+
 			// collect all sinks and add self loops to them
```
Comment (danieldietsch) [inline trunk/source/BuchiProgramProduct/src/de/uni_freiburg/informatik/ultimate/buchiprogramproduct/productgenerator/ProductGenerator.java]: update comment as well

File: trunk/source/BuchiProgramProduct/src/de/uni_freiburg/informatik/ultimate/buchiprogramproduct/productgenerator/ProductGenerator.java [hunk 8]
```diff
@@ -453,23 +465,20 @@ private void pruneNonProductSinks() {
 		// all helper states that have edges to non-product states receive a
 		// self-loop that keeps them in the same product state as the LTL NWA
 
-		for (final BoogieIcfgLocation helper : mHelperProductStates) {
+		for (final BoogieIcfgLocation helper : new HashSet<>(mHelperProductStates)) {
```
Comment (danieldietsch) [inline trunk/source/BuchiProgramProduct/src/de/uni_freiburg/informatik/ultimate/buchiprogramproduct/productgenerator/ProductGenerator.java]: does `pruneNonProductSink` generate new helper product states or why this `HashSet` here?

File: trunk/source/BuchiProgramProduct/src/de/uni_freiburg/informatik/ultimate/buchiprogramproduct/productgenerator/ProductGenerator.java [hunk 9]
```diff
@@ -488,36 +497,11 @@ private void pruneNonProductSink(final BoogieIcfgLocation helper) {
...
-		}
 		// hacky shit: the ss is now useless; we remove it
-		if (added) {
-			seq.disconnectSource();
-			seq.disconnectTarget();
-		}
+		//		if (added) {
+		seq.disconnectSource();
+		seq.disconnectTarget();
+		//		}
```
Comment (danieldietsch) [inline trunk/source/BuchiProgramProduct/src/de/uni_freiburg/informatik/ultimate/buchiprogramproduct/productgenerator/ProductGenerator.java]: If you remove all the stuff (491 to 515) you dont need to create a self loop and remove it again ;)

**Review by danieldietsch — CHANGES_REQUESTED** [review]

File: trunk/examples/LTL/regression/c/vanilla/Bug_ProcedureHandling.c (hunk 3 above)
Comment (Langenfeld) [reply trunk/examples/LTL/regression/c/vanilla/Bug_ProcedureHandling.c:2]: C does not allow the initialisation of global variables with a non-constant i.e. assigning nondet_int is not valid.
Change in the LTL property to accomodate the later initialisation.

**Review by danieldietsch — APPROVED** [review]
